### PR TITLE
Add dbuf hash and dbuf cache kstats

### PR DESCRIPTION
**Pull request by dinatale2**
### Description

Introduce kstats about the dbuf hash and dbuf cache
to make it easier to inspect state. This should help
with debugging and understanding of these portions
of the codebase.

Correct format of dbuf kstat file.

Introduce a dbc column to dbufs kstat to indicate if
a dbuf is in the dbuf cache.

Introduce field filtering in the dbufstat python script.

I will also be introducing some basic test cases to test the new dbufstats kstat and other basic scenarios.

### Motivation and Context

Gain a better understanding how dbufs are cached and provide another useful tool for users/developer.

### How Has This Been Tested?

Locally on a VM.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the ZFS on Linux code style requirements.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain `Signed-off-by`.
- [ ] Change has been approved by a ZFS on Linux member.

### CODE REVIEW

File: TEST [hunk 1]
```diff
@@ -5,25 +5,25 @@
 #TEST_PREPARE_SHARES="yes"
 
 ### SPLAT
-#TEST_SPLAT_SKIP="yes"
+TEST_SPLAT_SKIP="yes"
```
Comment (richardelling) [inline TEST]: are the changes to TEST appropriate for everyone?  methinks maybe this is a temporary change and not intended for broad adoption?

Comment (dinatale2) [reply TEST]: @richardelling These changes aren't going to be permanent. I just wanted to restrict the test suite to my new tests. Trying to figure out what I forgot to do to allow `test-runner` to accept my test.

File: module/zfs/dbuf.c [hunk 2]
```diff
@@ -48,6 +48,71 @@
...
+	{ "hash_chains",			KSTAT_DATA_UINT64 },
+	{ "hash_chain_max",			KSTAT_DATA_UINT64 },
+	{ "hash_insert_race",			KSTAT_DATA_UINT64 },
+	{ { "hash_dbuf_levels_N",		KSTAT_DATA_UINT64 } },
+	{ { "hash_dbuf_levels_bytes_N",		KSTAT_DATA_UINT64 } }
+};
+
+#define	DBUF_STAT_SET(stat, val)	\
+	(dbuf_stats.stat.value.ui64 = (val));
+#define	DBUF_STAT_INCR(stat, val)	\
```
Comment (tonyhutter) [inline module/zfs/dbuf.c]: Why not just the `atomic_*` macros directly?

File: module/zfs/dbuf.c [hunk 3]
```diff
@@ -48,6 +48,71 @@
...
+	atomic_add_64(&dbuf_stats.stat.value.ui64, (val));
+#define	DBUF_STAT_DECR(stat, val)	\
+	DBUF_STAT_INCR(stat, -(val));
+#define	DBUF_STAT_BUMP(stat)		\
+	DBUF_STAT_INCR(stat, 1);
+#define	DBUF_STAT_BUMPDOWN(stat)	\
+	DBUF_STAT_INCR(stat, -1);
+#define	DBUF_STAT_MAX(stat, val) {					\
```
Comment (tonyhutter) [inline module/zfs/dbuf.c]: Please make DBUF_STAT_MAX into a function since you're declaring variables.

File: module/zfs/dbuf.c [hunk 4]
```diff
@@ -687,6 +785,33 @@ dbuf_init(void)
 	cv_init(&dbuf_evict_cv, NULL, CV_DEFAULT, NULL);
 	dbuf_cache_evict_thread = thread_create(NULL, 0, dbuf_evict_thread,
 	    NULL, 0, &p0, TS_RUN, minclsyspri);
+
+	dbuf_ksp = kstat_create("zfs", 0, "dbuf_stats", "misc",
+	    KSTAT_TYPE_NAMED, sizeof (dbuf_stats) / sizeof (kstat_named_t),
+	    KSTAT_FLAG_VIRTUAL);
+	if (dbuf_ksp != NULL) {
```
Comment (tonyhutter) [inline module/zfs/dbuf.c]: You could do:
```
if (dbuf_ksp == NULL)
    return;
```
...and then you wouldn't need to indent the rest of the code below.

File: tests/zfs-tests/tests/functional/dbufs/cleanup.ksh [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+# file and include the License file at usr/src/OPENSOLARIS.LICENSE.
+# If applicable, add the following below this CDDL HEADER, with the
+# fields enclosed by brackets "[]" replaced with your own identifying
+# information: Portions Copyright [yyyy] [name of copyright owner]
+#
+# CDDL HEADER END
+#
+
+#
+# Copyright (c) 2016, Lawrence Livermore National Security, LLC.
```
Comment (tonyhutter) [inline tests/zfs-tests/tests/functional/dbufs/cleanup.ksh]: Should this be 2017 instead of 2016?

File: tests/zfs-tests/tests/functional/dbufs/dbufstats_001_pos.ksh [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+function testdbufstat # stat_name dbufstat_filter
+{
+        name=$1
+        filter=""
+
+        [[ -n "$2" ]] && filter="-F $2"
+
+        verify_eq \
+	    $(cat "$DBUFSTATS_FILE" | grep -w "$name" | awk '{ print $3 }') \
```
Comment (tonyhutter) [inline tests/zfs-tests/tests/functional/dbufs/dbufstats_001_pos.ksh]: Can you remove the `cat` and do `grep -w "$name" "$DBUFSTATS_FILE"` ?

File: tests/zfs-tests/tests/functional/dbufs/dbufstats_001_pos.ksh [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+#
+# STRATEGY:
+# 1. Generate a file with random data in it
+# 2. Store output from dbufs kstat
+# 3. Store output from dbuf_stats kstat
+# 4. Compare stats presented in dbuf_stats with stat generated using
+#    dbufstat.py and the dbufs kstat output
+#
+
+DBUFSTATS_FILE="/var/tmp/dbufstats.out.$$"
```
Comment (tonyhutter) [inline tests/zfs-tests/tests/functional/dbufs/dbufstats_001_pos.ksh]: Use `mktemp -d` to create a temporary directory instead of /var/tmp.

File: tests/zfs-tests/tests/functional/dbufs/dbufstats_001_pos.ksh [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+log_must dd if=/dev/urandom bs=1M count=20 of="$TESTDIR/file"
+log_must zpool sync
+
+sleep 5
+
+log_must eval "cat /proc/spl/kstat/zfs/dbufs > $DBUFS_FILE"
+log_must eval "cat /proc/spl/kstat/zfs/dbuf_stats > $DBUFSTATS_FILE"
+
+testdbufstat "dbuf_cache_count" "dbc=1"
+testdbufstat "dbuf_cache_level_0" "dbc=1,level=0"
```
Comment (tonyhutter) [inline tests/zfs-tests/tests/functional/dbufs/dbufstats_001_pos.ksh]: Can you run these tests in a for loop?
```
for level in {0..11} ; do
    testdbufstat "dbuf_cache_level_$level" "dbc=1,level=$level"
    testdbufstat "hash_dbuf_level_$level" "level=$level"
done
```

File: tests/zfs-tests/tests/functional/dbufs/dbufstats_002_pos.ksh [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+	log_must rm -f $TESTDIR/file $TESTDIR/file2 $DBUFS_FILE
+}
+
+verify_runnable "both"
+
+log_assert "dbufs move from mru to mfu list"
+
+log_onexit cleanup
+
+log_must dd if=/dev/urandom bs=1M count=1 of="$TESTDIR/file"
```
Comment (tonyhutter) [inline tests/zfs-tests/tests/functional/dbufs/dbufstats_002_pos.ksh]: You could make this more portable by using ZFS's `file_write` with the psudorandom option:
```file_write -o create -f "$TESTDIR/file" -b 1048576 -c 1 -d R```

File: tests/zfs-tests/tests/functional/cli_user/misc/dbufstat_001_pos.ksh [hunk 10]
```diff
@@ -37,4 +37,6 @@ while [[ $i -lt ${#args[*]} ]]; do
         ((i = i + 1))
 done
 
+log_must eval "dbufstat.py -F object=10,dbc=1,pool=testpool > /dev/null"
```
Comment (tonyhutter) [inline tests/zfs-tests/tests/functional/cli_user/misc/dbufstat_001_pos.ksh]: Can you add a quick comment describing what this does?  Also, should it be `pool=$TESTPOOL`?

File: cmd/dbufstat/dbufstat.py [hunk 11]
```diff
@@ -409,12 +414,28 @@ def update_dict(d, k, line, labels):
     return d
 
 
-def print_dict(d):
-    print_header()
+def skip_line(vals, filters):
```
Comment (tonyhutter) [inline cmd/dbufstat/dbufstat.py]: Can you add a quick comment describing what this function does?

File: module/zfs/dbuf.c (hunk 4 above)
Comment (dinatale2) [reply module/zfs/dbuf.c]: I thought about something like that, but I felt like having a block like that should be reserved for a condition that should be considered an "early exit" to a function (i.e. exits on errors). `dbuf_ksp` coming back as `NULL`, I don't believe, should be considered a terminal type condition.

File: module/zfs/dbuf.c (hunk 2 above)
Comment (dinatale2) [reply module/zfs/dbuf.c]: I did this to keep consistent with how `arcstats` is done.

File: module/zfs/dbuf.c (hunk 3 above)
Comment (dinatale2) [reply module/zfs/dbuf.c]: Renamed the variable to be `__m` so it's less likely to have a collision.

File: module/zfs/dbuf.c [hunk 12]
```diff
@@ -272,13 +337,15 @@ dbuf_hash_insert(dmu_buf_impl_t *db)
...
 	dmu_buf_impl_t *dbf;
+	uint64_t depth = 0;
 
 	blkid = db->db_blkid;
 	hv = dbuf_hash(os, obj, level, blkid);
 	idx = hv & h->hash_table_mask;
 
 	mutex_enter(DBUF_HASH_MUTEX(h, idx));
 	for (dbf = h->hash_table[idx]; dbf != NULL; dbf = dbf->db_hash_next) {
+		depth++;
```
Comment (behlendorf) [inline module/zfs/dbuf.c]: It'd be nice to model this for loop after the one in `buf_hash_insert()`.  It's fine to use `i` for the variable name here instead of `depth`.

File: module/zfs/dbuf.c [hunk 13]
```diff
@@ -289,11 +356,22 @@ dbuf_hash_insert(dmu_buf_impl_t *db)
 		}
 	}
 
+	if (depth > 0) {
+		DBUF_STAT_BUMP(hash_chains);
```
Comment (behlendorf) [inline module/zfs/dbuf.c]: This should bump `hash_collisions`.

File: module/zfs/dbuf.c [hunk 14]
```diff
@@ -289,11 +356,22 @@ dbuf_hash_insert(dmu_buf_impl_t *db)
...
+		if (depth == 1)
+			DBUF_STAT_BUMP(hash_collisions);
```
Comment (behlendorf) [inline module/zfs/dbuf.c]: This should bump `hash_chains`.

File: module/zfs/dbuf.c [hunk 15]
```diff
@@ -289,11 +356,22 @@ dbuf_hash_insert(dmu_buf_impl_t *db)
...
+		DBUF_STAT_MAX(hash_chain_max, depth);
+	}
+
 	mutex_enter(&db->db_mtx);
 	db->db_hash_next = h->hash_table[idx];
 	h->hash_table[idx] = db;
 	mutex_exit(DBUF_HASH_MUTEX(h, idx));
 	atomic_inc_64(&dbuf_hash_count);
+	DBUF_STAT_SET(hash_elements, dbuf_hash_count);
```
Comment (behlendorf) [inline module/zfs/dbuf.c]: Rather than always keep these two versions of this in sync I think it would make more sense to add a `ks_update` function.  Take a look at `arc_kstat_update` as an example.  This would allow you to remove entirely the `DBUF_STAT_SET` macro and simply dynamically generate these values when requested.

File: module/zfs/dbuf.c [hunk 16]
```diff
@@ -549,6 +643,12 @@ dbuf_evict_thread(void *unused)
 
 	mutex_enter(&dbuf_evict_lock);
 	while (!dbuf_evict_thread_exit) {
+		DBUF_STAT_SET(dbuf_cache_lowater_bytes,
+		    dbuf_cache_lowater_bytes());
+		DBUF_STAT_SET(dbuf_cache_hiwater_bytes,
+		    dbuf_cache_hiwater_bytes());
+		DBUF_STAT_SET(dbuf_cache_max_bytes, dbuf_cache_max_bytes);
```
Comment (behlendorf) [inline module/zfs/dbuf.c]: This all can be removed by adding a `ks_update` function.

File: module/zfs/dbuf.c [hunk 17]
```diff
@@ -48,6 +48,71 @@
...
+dbuf_stats_t dbuf_stats = {
+	{ "dbuf_cache_count",			KSTAT_DATA_UINT64 },
+	{ "dbuf_cache_size",			KSTAT_DATA_UINT64 },
+	{ "dbuf_cache_size_max",		KSTAT_DATA_UINT64 },
+	{ "dbuf_cache_max_bytes",		KSTAT_DATA_UINT64 },
+	{ "dbuf_cache_lowater_bytes",		KSTAT_DATA_UINT64 },
+	{ "dbuf_cache_hiwater_bytes",		KSTAT_DATA_UINT64 },
+	{ "dbuf_cache_total_evicts",		KSTAT_DATA_UINT64 },
+	{ { "dbuf_cache_levels_N",		KSTAT_DATA_UINT64 } },
+	{ { "dbuf_cache_levels_bytes_N",	KSTAT_DATA_UINT64 } },
```
Comment (behlendorf) [inline module/zfs/dbuf.c]: I find the `dbuf_` prefix here (and above) redundant, why don't you remove it the same way you did below with the `dbuf_hash`.   This way all the entries are consistent.

File: module/zfs/dbuf_stats.c [hunk 18]
```diff
@@ -46,14 +46,14 @@ static int
 dbuf_stats_hash_table_headers(char *buf, size_t size)
 {
 	(void) snprintf(buf, size,
-	    "%-88s | %-124s | %s\n"
-	    "%-16s %-8s %-8s %-8s %-8s %-8s %-8s %-5s %-5s %5s | "
-	    "%-5s %-5s %-8s %-6s %-8s %-12s "
-	    "%-6s %-6s %-6s %-6s %-6s %-8s %-8s %-8s %-5s | "
-	    "%-6s %-6s %-8s %-8s %-6s %-6s %-5s %-8s %-8s\n",
+	    "%-96s | %-119s | %s\n"
```
Comment (behlendorf) [inline module/zfs/dbuf_stats.c]: This was presumably to better align the existing output?

File: tests/zfs-tests/tests/functional/cli_user/misc/dbufstat_001_pos.ksh [hunk 19]
```diff
@@ -37,4 +37,7 @@ while [[ $i -lt ${#args[*]} ]]; do
         ((i = i + 1))
 done
 
+# a simple test of dbufstat.py filter functionality
```
Comment (behlendorf) [inline tests/zfs-tests/tests/functional/cli_user/misc/dbufstat_001_pos.ksh]: [nit] s/a/A/

File: tests/zfs-tests/tests/functional/dbufs/dbufs.kshlib [hunk 20]
```diff
@@ -0,0 +1,27 @@
...
+# information: Portions Copyright [yyyy] [name of copyright owner]
+#
+# CDDL HEADER END
+#
+
+#
+# Copyright (c) 2017, Lawrence Livermore National Security, LLC.
+#
+
+. $STF_SUITE/include/libtest.shlib
```
Comment (behlendorf) [inline tests/zfs-tests/tests/functional/dbufs/dbufs.kshlib]: Why add this file now if it doesn't do anything?

File: tests/zfs-tests/tests/functional/dbufs/dbufstats_001_pos.ksh [hunk 21]
```diff
@@ -0,0 +1,87 @@
...
+verify_runnable "both"
+
+log_assert "dbufstats produces correct statistics"
+
+log_onexit cleanup
+
+log_must file_write -o create -f "$TESTDIR/file" -b 1048576 -c 20 -d R
+log_must zpool sync
+
+sleep 5
```
Comment (behlendorf) [inline tests/zfs-tests/tests/functional/dbufs/dbufstats_001_pos.ksh]: Sleeping shouldn't be required, why was this needed?

File: module/zfs/dbuf_stats.c (hunk 18 above)
Comment (dinatale2) [reply module/zfs/dbuf_stats.c]: Yes it was.

File: tests/zfs-tests/tests/functional/dbufs/dbufstats_001_pos.ksh (hunk 21 above)
Comment (dinatale2) [reply tests/zfs-tests/tests/functional/dbufs/dbufstats_001_pos.ksh]: It's a remnant from before `zpool sync` was called. I'll remove it.

File: tests/zfs-tests/tests/functional/dbufs/dbufs.kshlib (hunk 20 above)
Comment (dinatale2) [reply tests/zfs-tests/tests/functional/dbufs/dbufs.kshlib]: I was thinking of moving the `testdbufstat" function into this file, but I don't think that needs to happen anymore. I'll get rid of this.

File: TEST (hunk 1 above)
Comment (behlendorf) [reply TEST]: When you next refresh this can you drop the changes to TEST so we can a full test run.

File: module/zfs/dbuf.c [hunk 22]
```diff
@@ -48,6 +48,96 @@
...
+	kstat_named_t cache_size;
+	kstat_named_t cache_size_max;
+	kstat_named_t cache_max_bytes;
+	/*
+	 * Statistics regarding the bounds on the dbuf cache size.
+	 */ 
+	kstat_named_t cache_target_bytes;
+	kstat_named_t cache_lowater_bytes;
+	kstat_named_t cache_hiwater_bytes;
+	kstat_named_t cache_total_evicts;
```
Comment (behlendorf) [inline module/zfs/dbuf.c]: Please add another comment for `cache_total_evicts`.

File: module/zfs/dbuf.c [hunk 23]
```diff
@@ -48,6 +48,96 @@
...
+	{ "hash_elements_max",			KSTAT_DATA_UINT64 },
+	{ "hash_chains",			KSTAT_DATA_UINT64 },
+	{ "hash_chains_max",			KSTAT_DATA_UINT64 },
+	{ "hash_insert_race",			KSTAT_DATA_UINT64 },
+	{ { "hash_dbuf_levels_N",		KSTAT_DATA_UINT64 } },
+	{ { "hash_dbuf_levels_bytes_N",		KSTAT_DATA_UINT64 } }
+};
+
+#define	DBUF_STAT_SET(stat, val)	\
+	(dbuf_stats.stat.value.ui64 = (val));
```
Comment (behlendorf) [inline module/zfs/dbuf.c]: `DBUF_STAT_SET` is unused and can be removed.

File: module/zfs/dbuf.c [hunk 24]
```diff
@@ -289,11 +381,22 @@ dbuf_hash_insert(dmu_buf_impl_t *db)
 		}
 	}
 
+	if (i > 0) {
+		DBUF_STAT_BUMP(hash_collisions);
+		if (i == 1)
+			DBUF_STAT_BUMP(hash_chains);
+
+		DBUF_STAT_MAX(hash_chains, i);
```
Comment (behlendorf) [inline module/zfs/dbuf.c]: You mean `hash_chains_max` here.

File: module/zfs/dbuf.c [hunk 25]
```diff
@@ -289,11 +381,22 @@ dbuf_hash_insert(dmu_buf_impl_t *db)
...
+	}
+
 	mutex_enter(&db->db_mtx);
 	db->db_hash_next = h->hash_table[idx];
 	h->hash_table[idx] = db;
 	mutex_exit(DBUF_HASH_MUTEX(h, idx));
 	atomic_inc_64(&dbuf_hash_count);
+	DBUF_STAT_MAX(hash_elements, dbuf_hash_count);
```
Comment (behlendorf) [inline module/zfs/dbuf.c]: You mean `hash_elements_max` here.

File: module/zfs/dbuf.c [hunk 26]
```diff
@@ -618,7 +737,30 @@ dbuf_evict_notify(void)
...
+		ds->cache_target_bytes.value.ui64 =
+		    dbuf_cache_target_bytes();
+		ds->cache_hiwater_bytes.value.ui64 =
+		    dbuf_cache_hiwater_bytes();
+		ds->cache_lowater_bytes.value.ui64 =
+		    dbuf_cache_lowater_bytes();
+		ds->cache_max_bytes.value.ui64 =
+		    dbuf_cache_max_bytes;
+		ds->hash_elements.value.ui64 =
+		    dbuf_hash_count;
```
Comment (behlendorf) [inline module/zfs/dbuf.c]: These all fit within the 80 character limit can you reformat them.

```
                ds->cache_size.value.ui64 = refcount_count(&dbuf_cache_size);
                ds->cache_target_bytes.value.ui64 = dbuf_cache_target_bytes();
                ds->cache_hiwater_bytes.value.ui64 = dbuf_cache_hiwater_bytes();
                ds->cache_lowater_bytes.value.ui64 = dbuf_cache_lowater_bytes();
                ds->cache_max_bytes.value.ui64 = dbuf_cache_max_bytes;
                ds->hash_elements.value.ui64 = dbuf_hash_count;
```

File: module/zfs/dbuf.c [hunk 27]
```diff
@@ -48,6 +48,96 @@
...
+kstat_t *dbuf_ksp;
+
+typedef struct dbuf_stats {
+	/*
+	 * Various statistics about the size of the dbuf cache.
+	 */ 
+	kstat_named_t cache_count;
+	kstat_named_t cache_size;
+	kstat_named_t cache_size_max;
+	kstat_named_t cache_max_bytes;
```
Comment (behlendorf) [inline module/zfs/dbuf.c]: I'd move `cache_max_bytes` after `cache_hiwater_bytes` below so they're grouped together.

File: module/zfs/dbuf.c [hunk 28]
```diff
@@ -48,6 +48,96 @@
...
 #include <sys/abd.h>
 
+kstat_t *dbuf_ksp;
+
+typedef struct dbuf_stats {
+	/*
+	 * Various statistics about the size of the dbuf cache.
+	 */ 
+	kstat_named_t cache_count;
+	kstat_named_t cache_size;
```
Comment (behlendorf) [inline module/zfs/dbuf.c]: Let's rename `cache_size` to `cache_size_bytes` to better fit in with the other `target`, `lowater`, `hiwater`, `max` values.

File: module/zfs/dbuf.c [hunk 29]
```diff
@@ -525,7 +637,14 @@ dbuf_evict_one(void)
 		multilist_sublist_unlock(mls);
 		(void) refcount_remove_many(&dbuf_cache_size,
 		    db->db.db_size, db);
+		DBUF_STAT_BUMPDOWN(cache_levels[db->db_level]);
+		DBUF_STAT_BUMPDOWN(cache_count);
+		DBUF_STAT_DECR(cache_levels_bytes[db->db_level],
+		    db->db.db_size);
 		dbuf_destroy(db);
+		DBUF_STAT_MAX(cache_size,
```
Comment (behlendorf) [inline module/zfs/dbuf.c]: You mean `cache_size_max` here.

File: module/zfs/dbuf.c [hunk 30]
```diff
@@ -3118,6 +3305,12 @@ dbuf_rele_and_unlock(dmu_buf_impl_t *db, void *tag)
 				multilist_insert(dbuf_cache, db);
 				(void) refcount_add_many(&dbuf_cache_size,
 				    db->db.db_size, db);
+				DBUF_STAT_BUMP(cache_levels[db->db_level]);
+				DBUF_STAT_BUMP(cache_count);
+				DBUF_STAT_INCR(cache_levels_bytes[db->db_level],
+				    db->db.db_size);
+				DBUF_STAT_MAX(cache_size,
```
Comment (behlendorf) [inline module/zfs/dbuf.c]: You mean `cache_size_max` here.

File: module/zfs/dbuf.c [hunk 31]
```diff
@@ -48,6 +48,96 @@
...
+	(dbuf_stats.stat.value.ui64 = (val));
+#define	DBUF_STAT_INCR(stat, val)	\
+	atomic_add_64(&dbuf_stats.stat.value.ui64, (val));
+#define	DBUF_STAT_DECR(stat, val)	\
+	DBUF_STAT_INCR(stat, -(val));
+#define	DBUF_STAT_BUMP(stat)		\
+	DBUF_STAT_INCR(stat, 1);
+#define	DBUF_STAT_BUMPDOWN(stat)	\
+	DBUF_STAT_INCR(stat, -1);
+#define	DBUF_STAT_MAX(stat, v) {					\
```
Comment (behlendorf) [inline module/zfs/dbuf.c]: Can you either rename this `DBUF_STAT_MAXSTAT` to match the arc, or drop the `_max` added by the macro and update the various call sites as indicated below.

File: module/zfs/dbuf.c [hunk 32]
```diff
@@ -48,6 +48,96 @@
 #include <sys/callb.h>
 #include <sys/abd.h>
 
+kstat_t *dbuf_ksp;
+
+typedef struct dbuf_stats {
+	/*
+	 * Various statistics about the size of the dbuf cache.
+	 */ 
+	kstat_named_t cache_count;
```
Comment (behlendorf) [inline module/zfs/dbuf.c]: Upon further reflection I don't think tracking the number of dbufs is actually all that useful.  Why don't we just keep the size in bytes and remove `cache_count` entirely.

File: module/zfs/dbuf.c (hunk 31 above)
Comment (dinatale2) [reply module/zfs/dbuf.c]: I ended up dropping the `_max` and made the changes you suggested below.

File: module/zfs/dbuf.c (hunk 32 above)
Comment (dinatale2) [reply module/zfs/dbuf.c]: I chose to keep it around.

File: tests/zfs-tests/tests/functional/dbufs/dbufstats_002_pos.ksh [hunk 33]
```diff
@@ -0,0 +1,86 @@
...
+# 8. Ensure that at least some dbufs moved to the mfu list in the ARC
+#
+# NOTES:
+# Currently, when dbufs are marked as DB_CACHED in the dbuf cache, the
+# ARC header for the dbuf is not updated to reflect the "hit". Therefore,
+# the only way to cause a dbuf to transition from mru to mfu is to ensure
+# that it is in the ARC, but not in the dbuf cache.
+#
+
+DBUFS_FILE="/var/tmp/dbufs.out.$$"
```
Comment (tonyhutter) [inline tests/zfs-tests/tests/functional/dbufs/dbufstats_002_pos.ksh]: Should use `mktemp` here

File: tests/zfs-tests/tests/functional/dbufs/dbufstats_002_pos.ksh [hunk 34]
```diff
@@ -0,0 +1,86 @@
...
+# 2. Generate a file with random data (small enough to fit in cache)
+# 3. zpool sync to remove dbufs from anon list in ARC
+# 4. Obtain the object ID using linux stat command
+# 5. Ensure that all dbufs are on the mru list in the ARC
+# 6. Generate another random file large enough to flush dbuf cache
+# 7. cat the first generated file
+# 8. Ensure that at least some dbufs moved to the mfu list in the ARC
+#
+# NOTES:
+# Currently, when dbufs are marked as DB_CACHED in the dbuf cache, the
```
Comment (dinatale2) [inline tests/zfs-tests/tests/functional/dbufs/dbufstats_002_pos.ksh]: Since this has been fixed, I need to remove the notes I added. I'll refresh this after the current test run.

File: module/zfs/dbuf.c [hunk 35]
```diff
@@ -469,28 +578,32 @@ dbuf_cache_target_bytes(void)
...
+}
 
-	uint64_t dbuf_cache_hiwater_bytes =
-	    (dbuf_cache_target * dbuf_cache_hiwater_pct) / 100;
+static inline uint64_t
+dbuf_cache_lowater_bytes(void)
+{
+	uint64_t dbuf_cache_target = dbuf_cache_target_bytes();
+	return (dbuf_cache_target -
+	    (dbuf_cache_target * dbuf_cache_hiwater_pct) / 100);
```
Comment (behlendorf) [inline module/zfs/dbuf.c]: This should be `dbuf_cache_lowater_bytes` not `dbuf_cache_hiwater_bytes`.

File: module/zfs/dbuf.c [hunk 36]
```diff
@@ -48,6 +48,97 @@
...
+	kstat_named_t cache_count;
+	kstat_named_t cache_size_bytes;
+	kstat_named_t cache_size_bytes_max;
+	/*
+	 * Statistics regarding the bounds on the dbuf cache size.
+	 */
+	kstat_named_t cache_target_bytes;
+	kstat_named_t cache_lowater_bytes;
+	kstat_named_t cache_hiwater_bytes;
+	kstat_named_t cache_max_bytes;
```
Comment (behlendorf) [inline module/zfs/dbuf.c]: Let's remove `cache_max_bytes` from the stats.  This value is already readable from the `dbuf_cache_max_bytes` module option so we're not losing any information.  But more importantly I think it's too easy to get confused with the `cache_size_bytes_max` since the names are almost identical.  Best to remove it and avoid any confusion entirely.

File: tests/runfiles/linux.run [hunk 37]
```diff
@@ -440,6 +440,10 @@ tags = ['functional', 'compression']
 tests = ['ctime_001_pos' ]
 tags = ['functional', 'ctime']
 
+[tests/functional/dbufs]
+tests = ['dbufstats_001_pos', 'dbufstats_002_pos']
+tags = ['functional', 'dbufs']
```
Comment (behlendorf) [inline tests/runfiles/linux.run]: What we want to be building up is a collection of test cases which verify that all aspects of the ARC are functioning as intended.  Since this covers far more than just dbufs can you rename this directory `tests/functional/arc`, this makes it clear where future ARC specific tests should be added.  There's no need to rename the new `dbufstats` test cases since they are testing a specific aspect of the ARC.

**Comment by behlendorf** [comment]
OK, generally this looks good but I did hit one issue while testing.  When running the ZTS it didn't take long before the `hash_dbuf_level_0_bytes` went negative.

```
hash_dbuf_level_0_bytes         4    18446744073671859712
```

**Comment by dinatale2** [comment]
I went ahead and removed the `hash_dbuf_level` stats because they really don't tell you much. I imagine the dbufs in the hash can also change in size which might explain why those stats were going negative occasionally.

**Comment by GregorKopka** [comment]
@behlendorf / @tonyhutter Can we get that into the next point release, please?
Would be very helpful when trying to look into issues like https://github.com/zfsonlinux/zfs/issues/7922

**Comment by behlendorf** [comment]
This would definitely be handy, but my guess it it would take a fair bit of effort to rework this code to apply to the next point release.
